src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs

```diff
@@ -747,31 +860,36 @@ public static MLDsa ImportSubjectPublicKeyInfo(ReadOnlySpan<byte> source)
                         AsnValueReader reader = new AsnValueReader(source, AsnEncodingRules.DER);
                         SubjectPublicKeyInfoAsn.Decode(ref reader, manager.Memory, out SubjectPublicKeyInfoAsn spki);
 
-                        MLDsaAlgorithm? algorithm = MLDsaAlgorithm.GetMLDsaAlgorithmFromOid(spki.Algorithm.Algorithm);
-
-                        if (algorithm is null)
-                        {
-                            throw Helpers.CreateAlgorithmUnknownException(spki.Algorithm.Algorithm);
-                        }
+                        MLDsaAlgorithm algorithm = GetAlgorithmIdentifier(ref spki.Algorithm);
+                        ReadOnlySpan<byte> publicKey = spki.SubjectPublicKey.Span;
 
-                        if (spki.Algorithm.Parameters.HasValue)
+                        if (publicKey.Length != algorithm.PublicKeySizeInBytes)
                         {
-                            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
-                            spki.Algorithm.Encode(writer);
-                            throw Helpers.CreateAlgorithmUnknownException(writer);
+                            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
                         }
 
-                        return MLDsaImplementation.ImportPublicKey(algorithm, spki.SubjectPublicKey.Span);
+                        return ImportMLDsaPublicKey(algorithm, spki.SubjectPublicKey.Span);
```

Comment:
Why switch this from calling the impl to calling the public method?  The public method is just going to recheck the length and revalidate the non-nullarity of `algorithm`.  (Unless the JIT sees through all that and turns it directly into the call to MLDsaImplementation.ImportPublicKey).